### PR TITLE
backlinks watcher accumulator

### PR DESCRIPTION
**Pull request by requilence**

### CODE REVIEW

File: core/block/source/source.go [hunk 1]
```diff
@@ -353,10 +360,10 @@ func (s *source) PushChange(params PushChangeParams) (id string, err error) {
 	if change.Snapshot != nil {
 		s.lastSnapshotId = id
 		s.changesSinceSnapshot = 0
-		log.Infof("%s: pushed snapshot", s.id)
+		log.Errorf("%s: pushed snapshot", s.id)
```
Comment (KirillSto) [inline core/block/source/source.go:363]: Level could be lowered back

File: core/block/source/source.go [hunk 2]
```diff
@@ -353,10 +360,10 @@ func (s *source) PushChange(params PushChangeParams) (id string, err error) {
...
 	} else {
 		s.changesSinceSnapshot++
-		log.Debugf("%s: pushed %d changes", s.id, len(change.Content))
+		log.Errorf("%s: pushed %d changes", s.id, len(change.Content))
```
Comment (KirillSto) [inline core/block/source/source.go:366]: Level could be lowered back

**Review by KirillSto — APPROVED** [review]

File: core/block/backlinks/watcher.go [hunk 3]
```diff
@@ -67,109 +75,169 @@ func (uw *UpdateWatcher) Close(context.Context) error {
...
+				process()
+				return
+			case <-time.After(aggregationInterval):
+				l.Lock()
+				if time.Since(lastReceivedUpdates) < aggregationInterval || len(accumulatedBacklinks) == 0 {
+					l.Unlock()
+					continue
+				}
+
+				log.Warnf("updating backlinks for %d objects", len(accumulatedBacklinks))
```
Comment (mcrakhman) [inline core/block/backlinks/watcher.go:150]: Fix here warn also pls

File: core/block/backlinks/watcher.go [hunk 4]
```diff
@@ -67,109 +75,169 @@ func (uw *UpdateWatcher) Close(context.Context) error {
...
+		l                    sync.Mutex
+		lastReceivedUpdates  time.Time
+		closedCh             = make(chan struct{})
+		aggregationInterval  = time.Second * 5
+	)
+	defer close(closedCh)
+
+	go func() {
+		process := func() {
+			log.Infof("updating backlinks for %d objects", len(accumulatedBacklinks))
```
Comment (mcrakhman) [inline core/block/backlinks/watcher.go:130]: It should be debug I guess

File: core/block/backlinks/watcher.go [hunk 5]
```diff
@@ -67,109 +75,169 @@ func (uw *UpdateWatcher) Close(context.Context) error {
...
-		for _, id := range modification.ids {
-			err = spc.DoLockedIfNotExists(id, func() error {
-				return uw.store.ModifyObjectDetails(id, modification.modifier)
+	err = spc.DoLockedIfNotExists(id, func() error {
+		return uw.store.ModifyObjectDetails(id, func(details *types.Struct) (*types.Struct, error) {
+			return updateBacklinks(details, backlinksUpdate)
+		})
+	})
+
+	if !errors.Is(err, ocache.ErrExists) {
```
Comment (mcrakhman) [inline core/block/backlinks/watcher.go:214]: If there is no error shouldn't you return from func, i.e. if you applied the changes when the object was not in cache?

Comment (KirillSto) [reply core/block/backlinks/watcher.go:218]: BTW yesterday I found many cases when error is DetailsNotChanged. That is absolutely harmless error, so should be considered as normal

**Pull request merged**